stan/math/rev/fun/sd.hpp

```diff
@@ -13,49 +13,93 @@
 namespace stan {
 namespace math {
 
+namespace internal {
 /**
  * Return the sample standard deviation of the specified std vector, column
- * vector, row vector, or matrix.
+ * vector, row vector, matrix, or std vector of any of these types.
  *
- * @tparam R number of rows, can be Eigen::Dynamic
- * @tparam C number of columns, can be Eigen::Dynamic
+ * @tparam T Input type
  * @param[in] m input matrix
+ * @return sample standard deviation
+ * @throw domain error  size is not greater than zero.
+ */
+template <typename T_map, require_st_var<T_map>* = nullptr,
+          require_not_var_matrix_t<T_map>* = nullptr>
+var sd_impl(const T_map& x) {
+  using std::sqrt;
+  using T_vi = promote_scalar_t<vari*, T_map>;
+  using T_d = promote_scalar_t<double, T_map>;
+
+  if (x.size() == 1) {
+    return 0.0;
+  }
+
+  vari** varis
+      = ChainableStack::instance_->memalloc_.alloc_array<vari*>(x.size());
+  double* partials
+      = ChainableStack::instance_->memalloc_.alloc_array<double>(x.size());
+  Eigen::Map<T_vi> varis_map(varis, x.rows(), x.cols());
+  Eigen::Map<T_d> partials_map(partials, x.rows(), x.cols());
+
+  varis_map = x.vi();
+  T_d dtrs_val = x.val();
+  double mean = dtrs_val.mean();
+  T_d diff = dtrs_val.array() - mean;
+  double sum_of_squares = diff.squaredNorm();
+  double size_m1 = x.size() - 1;
+  double sd = sqrt(sum_of_squares / size_m1);
+
+  if (sum_of_squares < 1e-20) {
+    partials_map.fill(inv_sqrt(static_cast<double>(x.size())));
+  } else {
+    partials_map = diff.array() / (sd * size_m1);
+  }
+  return var(new stored_gradient_vari(sd, x.size(), varis, partials));
+}
```

Comment:
I think you can get rid of this specialization if you just do the other one like

```cpp
template <typename T, require_rev_matrix_t<T>* = nullptr>
var sd_impl(const T& x) {
  if (x.size() == 1) {
    return 0.0;
  }
  arena_t<T> arena_x = x;
  auto arena_diff = to_arena((arena_x.val().array() - arena_x.val().mean()).matrix());
  double sum_of_squares = arena_diff.squaredNorm();
  double sd = std::sqrt(sum_of_squares / (x.size() - 1));

  var res = sd;

  reverse_pass_callback([arena_x, res, arena_diff]() mutable {
    arena_x.adj() += (res.adj() / (res.val() * (arena_x.size() - 1))) * arena_diff;
  });

  return res;
}
```

Also the requires for this impl makes it look like it will work for std vectors since it has `require_st_var` in it but it also uses `x.vi()` `x.val()` etc so idt it would work

Reply:
I want to avoid the `mat<var>` stuff cause that raises the question are we faster than what was there. We'll get to the benchmarking stuff eventually.

I should probably change this to `callback_vari` though now that I look at the code more closely.